src/mixins/themeable.ts

```diff
@@ -0,0 +1,154 @@
+			const currentClassNameFlag = oldCSSModuleClassNames[<keyof T> oldCSSModuleClassName];
+			// If it's true it needs to be negated and passed along, If it's false,
+			// don't return it as maquette will already have removed it.
+			if (currentClassNameFlag) {
+				newCSSModuleClassNames[oldCSSModuleClassName] = false;
+			}
+			return newCSSModuleClassNames;
+		}, <CSSModuleClassNames> {});
+
+		const calculatedClassNameMap = assign({}, negatedCSSModuleClassNames, newClasses[className]);
+		newAppliedClasses[className] = calculatedClassNameMap;
+
+		return newAppliedClasses;
+	}, <AppliedClasses<T>> {});
+}
+
+function generateThemeClasses<I, T>(instance: Themeable<I>, baseTheme: T, theme: {} = {}, overrideClasses: {} = {}) {
+	return Object.keys(baseTheme).reduce((newAppliedClasses, className: keyof T) => {
+		const newCSSModuleClassNames: CSSModuleClassNames = {};
+
+		let themeClassSource: {} = baseTheme;
+		if (theme.hasOwnProperty(className)) {
+			themeClassSource = theme;
+		}
+
+		addClassNameToCSSModuleClassNames(newCSSModuleClassNames, themeClassSource, className);
+		overrideClasses && addClassNameToCSSModuleClassNames(newCSSModuleClassNames, overrideClasses, className);
+		newAppliedClasses[className] = newCSSModuleClassNames;
+
+		return newAppliedClasses;
+	}, <AppliedClasses<T>> {});
+}
+
+function updateThemeClassesMap<I, T>(instance: Themeable<I>, newThemeClasses: AppliedClasses<T>) {
+	if (themeClassesMap.has(instance)) {
+		const previousThemeClasses = themeClassesMap.get(instance);
+		themeClassesMap.set(instance, negatePreviousClasses(previousThemeClasses, newThemeClasses));
+	} else {
+		themeClassesMap.set(instance, newThemeClasses);
+	}
+}
+
+function onPropertiesChanged<I>(instance: Themeable<I>, { theme, overrideClasses }: ThemeableProperties, changedPropertyKeys: string[]) {
+	const { theme: propTheme, overrideClasses: propOverrideClasses } = instance.properties;
+	const themeChanged = includes(changedPropertyKeys, 'theme');
+	const overrideClassesChanged = includes(changedPropertyKeys, 'overrideClasses');
+
+	if (themeChanged || overrideClassesChanged) {
+		const themeClasses = generateThemeClasses(instance, instance.baseTheme, theme || propTheme, overrideClasses || propOverrideClasses);
+		updateThemeClassesMap(instance, themeClasses);
+	}
+}
+
+/**
+ * Themeable Factory
+ */
+const themeableFactory: ThemeableFactory = createEvented.mixin({
+	mixin: {
+		get theme(this: Themeable<any>): AppliedClasses<any> {
+			return themeClassesMap.get(this);
+		}
+	},
+	initialize<I>(instance: Themeable<I>) {
+		instance.own(instance.on('properties:changed', (evt: PropertiesChangeEvent<ThemeableMixin<I>, ThemeableProperties>) => {
+			onPropertiesChanged(instance, evt.properties, evt.changedPropertyKeys);
+		}));
+		const { theme, overrideClasses } = instance.properties;
+		const themeClasses = generateThemeClasses(instance, instance.baseTheme, theme, overrideClasses);
```

Comment:
I just would prefer it if possible the initialise actually just called the same function as when the event was fired.